File: backend/app/schemas/audit.py
"""Public unified audit contracts."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class AuditEventResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    occurred_at: datetime
    company_id: UUID
    actor_user_id: UUID | None
    actor_role: str | None
    action: str
    resource_type: str
    resource_id: str | None
    result: str
    request_id: str | None
    correlation_id: str | None
    source_ip: str | None
    user_agent: str | None
    safe_metadata: dict[str, object]
    before_summary: str | None
    after_summary: str | None
    retention_class: str


class AuditEventPage(BaseModel):
    items: list[AuditEventResponse]
    total: int
    limit: int
    offset: int

File: frontend/src/api/audit.ts
import { apiDownload, apiRequest } from "./client";

export interface AuditEvent {
  readonly action: string;
  readonly actor_role: string | null;
  readonly actor_user_id: string | null;
  readonly after_summary: string | null;
  readonly before_summary: string | null;
  readonly correlation_id: string | null;
  readonly id: string;
  readonly occurred_at: string;
  readonly request_id: string | null;
  readonly resource_id: string | null;
  readonly resource_type: string;
  readonly result: "success" | "failure";
  readonly retention_class: string;
  readonly safe_metadata: Readonly<Record<string, unknown>>;
  readonly source_ip: string | null;
  readonly user_agent: string | null;
}

export interface AuditPage {
  readonly items: readonly AuditEvent[];
  readonly limit: number;
  readonly offset: number;
  readonly total: number;
}

export function listAuditEvents(query: {
  readonly action?: string;
  readonly offset: number;
  readonly resourceType?: string;
  readonly result?: "success" | "failure";
  readonly signal?: AbortSignal;
}): Promise<AuditPage> {
  const params = new URLSearchParams({
    limit: "50",
    offset: String(query.offset),
  });
  if (query.action) params.set("action", query.action);
  if (query.resourceType) params.set("resource_type", query.resourceType);
  if (query.result) params.set("result", query.result);
  return apiRequest<AuditPage>(`/audit-events?${params}`, {
    signal: query.signal,
  });
}

export async function downloadAuditEvents(format: "csv" | "json"): Promise<void> {
  const blob = await apiDownload(`/audit-events/export?export_format=${format}`);
  const url = URL.createObjectURL(blob);
  const link = document.createElement("a");
  link.href = url;
  link.download = `audit-events.${format}`;
  link.click();
  URL.revokeObjectURL(url);
}
